### PR TITLE
Pangle new API

### PR DESCRIPTION
**Pull request by wangyuanyuanyy**
Pangle new API adaptation

### CODE REVIEW

File: adapters/Pangle/Public/Headers/GADPangleNetworkExtras.h [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#import <Foundation/Foundation.h>
+
+NS_ASSUME_NONNULL_BEGIN
```
Comment (joshh-devrel) [inline adapters/Pangle/Public/Headers/GADPangleNetworkExtras.h:17]: Remove this.

File: adapters/Pangle/Public/Headers/GADPangleNetworkExtras.h [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+@interface GADPangleNetworkExtras : NSObject
+
+/// Set additional user information.
+@property (nonatomic, copy) NSString *userDataString;
+
+@end
+
+NS_ASSUME_NONNULL_END
```
Comment (joshh-devrel) [inline adapters/Pangle/Public/Headers/GADPangleNetworkExtras.h:26]: Remove this.

File: adapters/Pangle/Public/Headers/GADMediationAdapterPangle.h [hunk 3]
```diff
@@ -27,17 +27,18 @@ typedef NS_ENUM(NSInteger, GADPangleErrorCode) {
...
-/// "personal information" under CCPA. 0 means the user opts in. 1 means the user opts out. -1 means
-/// the user hasn't specified. Any value outside of -1, 0, or 1 will result in this method
+/// "personal information" under CCPA. 0 means “sale” of personal information is permitted.
+/// 1 means the user has opted out of “sale” of personal information.
+/// -1 means the user hasn't specified. Any value outside of -1, 0, or 1 will result in this method
 /// being a no-op.
-+ (void)setCCPA:(NSInteger)CCPA;
++ (void)doNotSell:(NSInteger)doNotSell;
 
 @end
```
Comment (joshh-devrel) [inline adapters/Pangle/Public/Headers/GADMediationAdapterPangle.h:44]: Can you update the comments according to the Android one?
https://github.com/googleads/googleads-mobile-android-mediation/pull/416/files#diff-f0b028dfe4ea0dedaebc67a0fa2655dae65547ff63fb237b7d246b4c891dda76L226-L265

**Comment by joshh-devrel** [comment]
@wangyuanyuanyy Can you please rebase the PR and fix the conflicts?

File: adapters/Pangle/PangleAdapter/GADMediationAdapterPangleConstants.h [hunk 4]
```diff
@@ -15,7 +15,7 @@
 #import <Foundation/Foundation.h>
 
 /// Pangle adapter version.
-static NSString *const GADMAdapterPangleVersion = @"4.3.1.9.0";
+static NSString *const GADMAdapterPangleVersion = @"4.6.1.8.0";
```
[the remark this answers is not in the document]
Comment (wangyuanyuanyy) [reply adapters/Pangle/PangleAdapter/GADMediationAdapterPangleConstants.h]: YES, I will fix it

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBBannerRenderer.m [hunk 5]
```diff
@@ -66,38 +61,60 @@ - (void)renderBannerForAdConfiguration:(nonnull GADMediationBannerAdConfiguratio
...
+    frame.size = bannerSize.size;
+    bannerAd.bannerView.frame = frame;
+    bannerAd.rootViewController = adConfiguration.topViewController;
+    
+    strongSelf->_bannerAd = bannerAd;
+    strongSelf->_bannerAd.delegate = strongSelf;
+    
+    if (strongSelf->_loadCompletionHandler) {
+      strongSelf->_delegate = strongSelf->_loadCompletionHandler(strongSelf,nil);
+    }
```
Comment (joshh-devrel) [inline adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBBannerRenderer.m:95]: Question: What is the reason that we need `strong` and `weak` here? What leads to this difference in the major release?

I don't see such implementation in Pangle documentation: https://www.pangleglobal.com/integration/ios-banner-ads.

Comment (joshh-devrel) [reply adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBBannerRenderer.m:95]: Seems like it's because of these lines:
```
    if (strongSelf->_loadCompletionHandler) {
      strongSelf->_delegate = strongSelf->_loadCompletionHandler(strongSelf,nil);
    }
```
But still it would be great if you can clarify that in Pangle SDK perspective.

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBInterstitialRenderer.m [hunk 6]
```diff
@@ -13,102 +13,93 @@
...
+      };
+
+    NSString *placementId = adConfiguration.credentials.settings[GADMAdapterPanglePlacementID];
+    if (!placementId.length) {
+      NSError *error = GADMAdapterPangleErrorWithCodeAndDescription(
+          GADPangleErrorInvalidServerParameters,
+          [NSString stringWithFormat:@"%@ cannot be nil.", GADMAdapterPanglePlacementID]);
+      _loadCompletionHandler(nil, error);
+      return;
+    }
```
Comment (joshh-devrel) [inline adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBInterstitialRenderer.m:59]: If there are not changes here, can you revert the formatting changes?

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m [hunk 7]
```diff
@@ -13,83 +13,99 @@
...
-  id<GADMediationNativeAdEventDelegate> _delegate;
+@interface GADPangleRTBNativeRenderer()<PAGLNativeAdDelegate> {
+    /// The completion handler to call when the ad loading succeeds or fails.
+    GADMediationNativeLoadCompletionHandler _loadCompletionHandler;
+    /// The Pangle native ad.
+    PAGLNativeAd *_nativeAd;
+    /// The Pangle related view
+    PAGLNativeAdRelatedView *_relatedView;
+    /// An ad event delegate to invoke when ad rendering events occur.
+    id<GADMediationNativeAdEventDelegate> _delegate;
```
Comment (joshh-devrel) [inline adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m:29]: Revert all the indentation change.

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m [hunk 8]
```diff
@@ -13,83 +13,99 @@
...
+        return delegate;
+    };
+    
+    NSString *placementId = adConfiguration.credentials.settings[GADMAdapterPanglePlacementID] ?: @"";
+    if (!placementId.length) {
+        NSError *error = GADMAdapterPangleErrorWithCodeAndDescription(GADPangleErrorInvalidServerParameters,
+                                                                      [NSString stringWithFormat:@"%@ cannot be nil.",GADMAdapterPanglePlacementID]
+                                                                      );
+        _loadCompletionHandler(nil, error);
+        return;
```
Comment (joshh-devrel) [inline adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m:59]: Revert all the formatting change.

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m [hunk 9]
```diff
@@ -13,83 +13,99 @@
...
-  /// The Pangle related view.
-  BUNativeAdRelatedView *_relatedView;
-  /// An ad event delegate to invoke when ad rendering events occur.
-  id<GADMediationNativeAdEventDelegate> _delegate;
+@interface GADPangleRTBNativeRenderer()<PAGLNativeAdDelegate> {
+    /// The completion handler to call when the ad loading succeeds or fails.
+    GADMediationNativeLoadCompletionHandler _loadCompletionHandler;
+    /// The Pangle native ad.
+    PAGLNativeAd *_nativeAd;
+    /// The Pangle related view
```
Comment (joshh-devrel) [inline adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m:26]: Add the `.` back.

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m [hunk 10]
```diff
@@ -168,52 +172,32 @@ - (BOOL)handlesUserImpressions {
...
-  if (_loadCompletionHandler) {
-    _loadCompletionHandler(nil, error);
-  }
+- (void)didRenderInView:(nonnull UIView *)view
+       clickableAssetViews:
+           (nonnull NSDictionary<GADNativeAssetIdentifier, UIView *> *)clickableAssetViews
+    nonclickableAssetViews:
+        (nonnull NSDictionary<GADNativeAssetIdentifier, UIView *> *)nonclickableAssetViews
+         viewController:(nonnull UIViewController *)viewController {
+    [_nativeAd registerContainer:view withClickableViews:clickableAssetViews.allValues];
```
Comment (joshh-devrel) [inline adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBNativeRenderer.m:188]: Seems like these lines didn't change except the indentation? Revert the formatting changes.

File: adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m [hunk 11]
```diff
@@ -13,32 +13,37 @@
...
-  /// Pangle native ad wrapper.
-  GADPangleRTBNativeRenderer *_nativeRenderer;
+    /// Pangle banner ad wrapper.
+    GADPangleRTBBannerRenderer *_bannerRenderer;
+    /// Pangle interstitial ad wrapper.
+    GADPangleRTBInterstitialRenderer *_interstitialRenderer;
+    /// Pangle rewarded ad wrapper.
+    GADPangleRTBRewardedRenderer *_rewardedRenderer;
+    /// Pangle native ad wrapper.
+    GADPangleRTBNativeRenderer *_nativeRenderer;
```
Comment (joshh-devrel) [inline adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m:35]: Revert the formatting changes.

File: adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m [hunk 12]
```diff
@@ -13,32 +13,37 @@
...
 - (void)collectSignalsForRequestParameters:(nonnull GADRTBRequestParameters *)params
                          completionHandler:
-                             (nonnull GADRTBSignalCompletionHandler)completionHandler {
-  NSString *signals = [BUAdSDKManager getBiddingToken:nil];
-  completionHandler(signals, nil);
+(nonnull GADRTBSignalCompletionHandler)completionHandler {
+    GADPangleNetworkExtras *extras = [params.extras isKindOfClass:[GADPangleNetworkExtras class]] ? params.extras : nil;
+    if (extras && extras.userDataString.length > 0) {
+        [PAGConfig shareConfig].userDataString = extras.userDataString;
+    }
```
Comment (joshh-devrel) [inline adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m:44]: Why do we need the extras for collecting signals?

File: adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m [hunk 13]
```diff
@@ -142,51 +145,49 @@ - (void)loadRewardedAdForAdConfiguration:(GADMediationRewardedAdConfiguration *)
...
-/// child-directed for purposes of the COPPA.  0 means false. 1 means true. -1 means
-/// unspecified. Any value outside of -1, 0, or 1 will result in this method being a no-op.
-+ (void)setCOPPA:(NSInteger)COPPA {
-  if (COPPA != 0 && COPPA != 1 && COPPA != -1) {
+/// @param childDirected An integer value that indicates whether the app should be treated as
+/// child-directed for purposes of the COPPA.  See <a
+/// href="https://www.pangleglobal.com/integration/ios-initialize-pangle-sdk">
+/// Pangle's documentation</a> for more information about what values may be provided.
++ (void)setChildDirected:(NSInteger)childDirected {
+  if (childDirected != 0 && childDirected != 1 && childDirected != -1) {
```
Comment (joshh-devrel) [inline adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m:163]: Revert this change.

File: adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m [hunk 14]
```diff
@@ -142,51 +145,49 @@ - (void)loadRewardedAdForAdConfiguration:(GADMediationRewardedAdConfiguration *)
...
     GADMPangleLog(@"Invalid COPPA value. Pangle SDK only accepts -1, 0 or 1.");
```
Comment (joshh-devrel) [inline adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m:164]: Does Pangle SDK support `-1`? I only saw the below values in Pangle doc:
```
COPPA
PAGChildDirectedTypeNonChild	:adult
PAGChildDirectedTypeChild	:child
```

File: adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m [hunk 15]
```diff
@@ -142,51 +145,49 @@ - (void)loadRewardedAdForAdConfiguration:(GADMediationRewardedAdConfiguration *)
...
     return;
   }
-  if (BUAdSDKManager.initializationState == BUAdSDKInitializationStateReady) {
-    [BUAdSDKManager setCoppa:COPPA];
+  if (PAGSdk.initializationState == PAGSDKInitializationStateReady) {
+      PAGConfig.shareConfig.childDirected = childDirected;
```
Comment (joshh-devrel) [inline adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m:168]: We should evaluate the input `COPPA` value before passing to Pangle SDK.

Note that:
- `YES` (or `1`) means the content will be treated as child-directed for purposes of COPPA.
    This should be equivalent to `PAGChildDirectedTypeChild`, but is that `1`?
- `NO` (or `0`) means the content won't be treated as child-directed for purposes of COPPA.
    This should be equivalent to `PAGChildDirectedTypeNonChild `, but is that `0`?

File: adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m [hunk 16]
```diff
@@ -142,51 +145,49 @@ - (void)loadRewardedAdForAdConfiguration:(GADMediationRewardedAdConfiguration *)
...
   }
 }
-
-+ (void)setGDPR:(NSInteger)GDPR {
-  if (GDPR != 0 && GDPR != 1 && GDPR != -1) {
++ (void)GDPRConsent:(NSInteger)GDPRConsent {
```
Comment (joshh-devrel) [inline adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m:171]: Update the method name to `setGDPRConsent`.

File: adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m [hunk 17]
```diff
@@ -142,51 +145,49 @@ - (void)loadRewardedAdForAdConfiguration:(GADMediationRewardedAdConfiguration *)
...
+  if (PAGSdk.initializationState == PAGSDKInitializationStateReady) {
+      PAGConfig.shareConfig.GDPRConsent = GDPRConsent;
   }
-  _gdpr = GDPR;
+  _GDPRConsent = GDPRConsent;
 }
 
-+ (void)setCCPA:(NSInteger)CCPA {
-  if (CCPA != 0 && CCPA != 1 && CCPA != -1) {
++ (void)doNotSell:(NSInteger)doNotSell {
```
Comment (joshh-devrel) [inline adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m:182]: Update the method name to `setDoNotSell`.

File: adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m [hunk 18]
```diff
@@ -142,51 +145,49 @@ - (void)loadRewardedAdForAdConfiguration:(GADMediationRewardedAdConfiguration *)
...
+  if (doNotSell != 0 && doNotSell != 1 && doNotSell != -1) {
```
Comment (joshh-devrel) [inline adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m:183]: Only two values are supported, shouldn't be something like below?
```
  if (doNotSell != PAGGDPRConsentTypeNoConsent && doNotSell != PAGGDPRConsentTypeConsent) {
```

File: adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m [hunk 19]
```diff
@@ -142,51 +145,49 @@ - (void)loadRewardedAdForAdConfiguration:(GADMediationRewardedAdConfiguration *)
...
-    [BUAdSDKManager setCoppa:COPPA];
+  if (PAGSdk.initializationState == PAGSDKInitializationStateReady) {
+      PAGConfig.shareConfig.childDirected = childDirected;
   }
 }
-
-+ (void)setGDPR:(NSInteger)GDPR {
-  if (GDPR != 0 && GDPR != 1 && GDPR != -1) {
++ (void)GDPRConsent:(NSInteger)GDPRConsent {
+  if (GDPRConsent != 0 && GDPRConsent != 1 && GDPRConsent != -1) {
```
Comment (joshh-devrel) [inline adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m:172]: Only two values are supported, shouldn't be something like below?
```
  if (GDPRConsent != PAGGDPRConsentTypeNoConsent && GDPRConsent != PAGGDPRConsentTypeConsent) {
```

File: adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m [hunk 20]
```diff
@@ -64,19 +69,17 @@ + (void)setUpWithConfiguration:(nonnull GADMediationServerConfiguration *)config
...
-  BUAdSDKConfiguration *sdkConfiguration = [BUAdSDKConfiguration configuration];
-  sdkConfiguration.territory = BUAdSDKTerritory_NO_CN;
-  sdkConfiguration.appID = appID;
-  sdkConfiguration.GDPR = @(_gdpr);
-  sdkConfiguration.CCPA = @(_ccpa);
-  [BUAdSDKManager startWithAsyncCompletionHandler:^(BOOL success, NSError *error) {
+  PAGConfig *config = [PAGConfig shareConfig];
+  config.appID = appID;
+  config.GDPRConsent = _GDPRConsent;
+  config.doNotSell = _doNotSell;
```
Comment (joshh-devrel) [inline adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m:75]: Does Pangle SDK accept `-1` for the `GDPRConsent` or `doNotSell` value?

File: adapters/Pangle/Public/Headers/GADMediationAdapterPangle.h [hunk 21]
```diff
@@ -26,18 +26,18 @@ typedef NS_ENUM(NSInteger, GADPangleErrorCode) {
...
-/// @param GDPR An integer value that indicates whether the user consents the use of personal
-/// data to serve ads under GDPR.  0 means the user consents. 1 means the user does not consent.
-/// -1 means the user hasn't specified. Any value outside of -1, 0, or 1 will result in this method
-/// being a no-op.
-+ (void)setGDPR:(NSInteger)GDPR;
+/// @param GDPRConsent An integer value that indicates whether the user consents the use of personal
+/// data to serve ads under GDPR. See <a
+/// href="https://www.pangleglobal.com/integration/ios-initialize-pangle-sdk">
+/// Pangle's documentation</a> for more information about what values may be provided.
++ (void)GDPRConsent:(NSInteger)GDPRConsent;
```
Comment (joshh-devrel) [inline adapters/Pangle/Public/Headers/GADMediationAdapterPangle.h:33]: Update the method name to `setGDPRConsent`.

File: adapters/Pangle/Public/Headers/GADMediationAdapterPangle.h [hunk 22]
```diff
@@ -26,18 +26,18 @@ typedef NS_ENUM(NSInteger, GADPangleErrorCode) {
...
-/// @param CCPA An integer value that indicates whether the user opts in of the "sale" of the
-/// "personal information" under CCPA. 0 means the user opts in. 1 means the user opts out. -1 means
-/// the user hasn't specified. Any value outside of -1, 0, or 1 will result in this method
-/// being a no-op.
-+ (void)setCCPA:(NSInteger)CCPA;
+/// @param doNotSell  An integer value that indicates whether the user opts in of the "sale" of the
+/// "personal information" under CCPA. See <a
+/// href="https://www.pangleglobal.com/integration/ios-initialize-pangle-sdk">
+/// Pangle's documentation</a> for more information about what values may be provided.
++ (void)doNotSell:(NSInteger)doNotSell;
```
Comment (joshh-devrel) [inline adapters/Pangle/Public/Headers/GADMediationAdapterPangle.h:41]: Update the method name to `setDoNotSell`.

File: adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m (hunk 12 above)
Comment (joshh-devrel) [reply adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m:44]: When should the publishers set this user information? Should it be before Pangle SDK initializes or at ad request time?

Comment (wangyuanyuanyy) [reply adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m:44]: 1. Custom parameters for sending Pangle the user details, like in-app purchases status. To enable this function please contact your Pangle's AM for more information 2.Publishers set this user information at ad request time.

File: adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m (hunk 13 above)
Comment (wangyuanyuanyy) [reply adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m:163]: Due to Pangle SDK changed property name.

File: adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m (hunk 14 above)
Comment (wangyuanyuanyy) [reply adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m:164]: In Pangle SDK PAGConfig Header 
`typedef NS_ENUM(NSInteger, PAGChildDirectedType) {
    PAGChildDirectedTypeDefault = -1,//default
    PAGChildDirectedTypeNonChild = 0,// user is not a child
    PAGChildDirectedTypeChild = 1,// user is a child
};`

File: adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m (hunk 15 above)
Comment (wangyuanyuanyy) [reply adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m:168]: Yes.

File: adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m (hunk 18 above)
Comment (wangyuanyuanyy) [reply adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m:183]: We support three values. default(-1), noConsent(-1), consent(-1)

File: adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m (hunk 19 above)
Comment (wangyuanyuanyy) [reply adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m:172]: We support three values. default(-1), noConsent(0), consent(1)

File: adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m (hunk 20 above)
Comment (wangyuanyuanyy) [reply adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m:75]: Yes, Pangle SDK accept -1 for GDPTConsent 、 doNotSell and childDirected.

File: adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBBannerRenderer.m (hunk 5 above)
Comment (wangyuanyuanyy) [reply adapters/Pangle/PangleAdapter/Bidding/GADPangleRTBBannerRenderer.m:95]: To prevent circular references

File: adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m (hunk 12 above)
Comment (joshh-devrel) [reply adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m:44]: @wangyuanyuanyy 
This shouldn't be set here, please remove.

Like you have said, this should be set at ad request time.
Please create a utils class and a method within it, then in `render()` call that method that set the user data.

See Android PR for example:
https://github.com/googleads/googleads-mobile-android-mediation/pull/416/files#diff-b814f329c3ba459244ae4a400ca481c4ab6cfba6d8fc9341e05e90c1b0f3514eR41

Comment (wangyuanyuanyy) [reply adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m:44]: Sorry, I misunderstood what you meant. Publishers set this user information before get bidding token.

Comment (joshh-devrel) [reply adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m:44]: @wangyuanyuanyy 

Follow-up(s):
- Does `[PAGSdk getBiddingToken:nil]` use that user data set in `[PAGConfig shareConfig].userDataString` to generate token (or signal)?

File: adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m (hunk 13 above)
Comment (joshh-devrel) [reply adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m:163]: This method will only be used by the adapter, not publisher, so we don't need its name matches the one in Pangle SDK. Can you revert the change?

File: adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m [hunk 23]
```diff
@@ -36,9 +37,13 @@ @implementation GADMediationAdapterPangle {
 
 - (void)collectSignalsForRequestParameters:(nonnull GADRTBRequestParameters *)params
                          completionHandler:
-                             (nonnull GADRTBSignalCompletionHandler)completionHandler {
-  NSString *signals = [BUAdSDKManager getBiddingToken:nil];
-  completionHandler(signals, nil);
+(nonnull GADRTBSignalCompletionHandler)completionHandler {
+    GADPangleNetworkExtras *extras = [params.extras isKindOfClass:[GADPangleNetworkExtras class]] ? params.extras : nil;
+    if (extras && extras.userDataString.length > 0) {
+        [PAGConfig shareConfig].userDataString = extras.userDataString;
```
Comment (joshh-devrel) [inline adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m:43]: Add this comment:
```
// The user data needs to be set for it to be included in the signals.
```

File: adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m (hunk 12 above)
Comment (wangyuanyuanyy) [reply adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m:44]: Yes,`[PAGSdk getBiddingToken:nil]`user that user data set in `[PAGConfig shareConfig].userDataString` to get generate token.

File: adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m (hunk 13 above)
Comment (wangyuanyuanyy) [reply adapters/Pangle/PangleAdapter/GADMediationAdapterPangle.m:163]: ok~
